ethcore/src/snapshot/service.rs

```diff
@@ -406,55 +416,133 @@ impl Service {
+		if let Err(e) = fs::remove_dir_all(&prev_chunks) {
 			match e.kind() {
 				ErrorKind::NotFound => {},
 				_ => return Err(e.into()),
 			}
 		}
 
-		fs::create_dir_all(&rest_dir)?;
+		// Move the previous recovery temp directory
+		// to `prev_chunks` to be able to restart restoring
+		// with previously downloaded blocks
+		// This step is optional, so don't fail on error
+		fs::rename(&recovery_temp, &prev_chunks).ok();
 
-		// make new restoration.
-		let writer = match recover {
-			true => Some(LooseWriter::new(self.temp_recovery_dir())?),
-			false => None
-		};
+		{
+			let mut res = self.restoration.lock();
 
-		let params = RestorationParams {
-			manifest: manifest,
-			pruning: self.pruning,
-			db: self.restoration_db_handler.open(&self.restoration_db())?,
-			writer: writer,
-			genesis: &self.genesis_block,
-			guard: Guard::new(rest_dir),
-			engine: &*self.engine,
-		};
+			self.state_chunks.store(0, Ordering::SeqCst);
+			self.block_chunks.store(0, Ordering::SeqCst);
 
-		let state_chunks = params.manifest.state_hashes.len();
-		let block_chunks = params.manifest.block_hashes.len();
+			// tear down existing restoration.
+			*res = None;
 
-		*res = Some(Restoration::new(params)?);
+			// delete and restore the restoration dir.
+			if let Err(e) = fs::remove_dir_all(&rest_dir) {
+				match e.kind() {
+					ErrorKind::NotFound => {},
+					_ => return Err(e.into()),
+				}
+			}
 
-		*self.status.lock() = RestorationStatus::Ongoing {
-			state_chunks: state_chunks as u32,
-			block_chunks: block_chunks as u32,
-			state_chunks_done: self.state_chunks.load(Ordering::SeqCst) as u32,
-			block_chunks_done: self.block_chunks.load(Ordering::SeqCst) as u32,
-		};
+			fs::create_dir_all(&rest_dir)?;
+
+			// make new restoration.
+			let writer = match recover {
+				true => Some(LooseWriter::new(self.temp_recovery_dir())?),
+				false => None
+			};
+
+			let params = RestorationParams {
+				manifest: manifest.clone(),
+				pruning: self.pruning,
+				db: self.restoration_db_handler.open(&self.restoration_db())?,
+				writer: writer,
+				genesis: &self.genesis_block,
+				guard: Guard::new(rest_db),
+				engine: &*self.engine,
+			};
+
+			let state_chunks = manifest.state_hashes.len();
+			let block_chunks = manifest.block_hashes.len();
+
+			*res = Some(Restoration::new(params)?);
+
+			*self.status.lock() = RestorationStatus::Ongoing {
+				state_chunks: state_chunks as u32,
+				block_chunks: block_chunks as u32,
+				state_chunks_done: self.state_chunks.load(Ordering::SeqCst) as u32,
+				block_chunks_done: self.block_chunks.load(Ordering::SeqCst) as u32,
+			};
+		}
 
 		self.restoring_snapshot.store(true, Ordering::SeqCst);
+
+		// Import previous chunks, continue if it fails
+		self.import_prev_chunks(manifest).ok();
+		self.ready.store(true, Ordering::SeqCst);
+
+		Ok(())
+	}
+
+	/// Import the previous chunks into the current restoration
+	fn import_prev_chunks(&self, manifest: ManifestData) -> Result<(), Error> {
```

Comment:
Isn't there a better way to figure out if the chunks in directory are really mean for this manifest?

Reply:
We could check for the name first, but hashing the content of the chunks is really fast so I don't think it's a bottleneck here (and it shouldn't happen that much that you stop your warp sync to restart syncing another snapshot)